src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreePostingsWriter.java

```diff
@@ -0,0 +1,198 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Writes auxiliary posting lists for bbtree nodes. If a node has a posting list attached, it will contain every row
+ * id from all leaves reachable from that node.
+ * <p>
+ * Writer is stateful, because it needs to collect data from the balanced tree data structure first to find set of eligible
+ * nodes and leaf nodes reachable from them.
+ * <p>
+ * This is an optimised writer for 1-dim points, where we know that leaf blocks are written in value order (in this
+ * order we pass them to the {@link BlockBalancedTreeWriter}). That allows us to skip reading the leaves, instead
+ * just order leaf blocks by their offset in the index file, and correlate them with buffered posting lists.
+ */
+public class BlockBalancedTreePostingsWriter implements TraversingBlockBalancedTreeReader.IndexTreeTraversalCallback
+{
+    private static final Logger logger = LoggerFactory.getLogger(BlockBalancedTreePostingsWriter.class);
+
+    /**
+     * Minimum number of reachable leaves for a given node to be eligible for an auxiliary posting list.
+     */
+    public static final int MINIMUM_POSTINGS_LEAVES = CassandraRelevantProperties.SAI_MINIMUM_POSTINGS_LEAVES.getInt();
+    /**
+     * Skip, or the sampling interval, for selecting a balanced tree level that is eligible for an auxiliary posting list.
+     * Sampling starts from 0, but the balanced tree root node is at level 1. For skip = 4, eligible levels are 4, 8, 12, etc. (no
+     * level 0, because there is no node at level 0).
+     */
+    public static final int POSTINGS_SKIP = CassandraRelevantProperties.SAI_POSTINGS_SKIP.getInt();
+
+    private final List<PackedLongValues> postings;
+    private final TreeMap<Long, Integer> leafOffsetToNodeID = new TreeMap<>(Long::compareTo);
+    private final Multimap<Integer, Integer> nodeToChildLeaves = HashMultimap.create();
+
+    private final IndexContext indexContext;
+    int numNonLeafPostings = 0;
+    int numLeafPostings = 0;
+
+    BlockBalancedTreePostingsWriter(List<PackedLongValues> postings, IndexContext indexContext)
+    {
+        this.postings = postings;
+        this.indexContext = indexContext;
+    }
+
+    @Override
+    public void onLeaf(int leafNodeID, long leafBlockFP, IntArrayList pathToRoot)
+    {
+        checkArgument(!pathToRoot.containsInt(leafNodeID));
+        checkArgument(pathToRoot.isEmpty() || leafNodeID > pathToRoot.get(pathToRoot.size() - 1));
+
+        leafOffsetToNodeID.put(leafBlockFP, leafNodeID);
+        for (int i = 0; i < pathToRoot.size(); i++)
+        {
+            int level = i + 1;
+            if (isLevelEligibleForPostingList(level))
+            {
+                int nodeID = pathToRoot.get(i);
+                nodeToChildLeaves.put(nodeID, leafNodeID);
+            }
+        }
+    }
+
+    public long finish(IndexOutput out) throws IOException
```

Comment:
Feels like a JavaDoc on the method would be helpful, if only to be clear about what the returned position actually is. (It looks like the position in the postings file where the node ID -> posting list map is written, rather than the footer?)

Reply:
I have done this and have noted a few optimisations that can be made to this process.